xarray/core/indexing.py

```diff
@@ -654,6 +686,185 @@ def _outer_to_numpy_indexer(key, shape):
+    if isinstance(indexer, BasicIndexer):
+        return indexer, ()
+    raise TypeError('unexpected key type: {}'.format(indexer))
+
+
+def _decompose_vectorized_indexer(indexer, shape, mode):
+    """
+    Decompose vectorized indexer to the successive two indexers, where the
+    first indexer will be used to index backend arrays, while the second one
+    is used to index loaded on-memory np.ndarray.
+
+    Parameters
+    ----------
+    indexer: VectorizedIndexer
+    mode: str, one of ['basic' | 'outer' | 'outer_1vector' | 'vectorized']
+        basic: for backends that support onle basic indexer
+        outer: for backends that support basic / outer indexer
+        outer_1vector: for backends that support outer indexer inluding at most
+                       1 vector.
+        vectorized: for backends that support full vectorized indexer.
+
+    Returns
+    -------
+    backend_indexer: OuterIndexer or BasicIndexer
+    np_indexers: a sequence of vectorizedIndexer
+
+    Note
+    ----
+    This function is used to realize the vectorized indexing for the backend
+    arrays that only support basic or outer indexing.
+
+    As an example, let us consider to index a few elements from a backend array
+    with a vectorized indexer ([0, 3, 1], [2, 3, 2]).
+    Even if the backend array only supports outer indexing, it is more
+    efficient to load a subslice of the array than loading the entire array,
+
+    >>> backend_indexer = OuterIndexer([0, 1, 3], [2, 3])
+    >>> array = array[backend_indexer]  # load subslice of the array
+    >>> np_indexer = VectorizedIndexer([0, 2, 1], [0, 1, 0])
+    >>> array[np_indexer]  # vectorized indexing for on-memory np.ndarray.
+    """
+    assert isinstance(indexer, VectorizedIndexer)
+
+    if mode == 'vectorized':
+        return indexer, ()
+
+    backend_indexer = []
+    np_indexer = []
+    # posify
+    indexer = [np.where(k < 0, k + s, k) if isinstance(k, np.ndarray) else k
+               for k, s in zip(indexer.tuple, shape)]
+
+    for k in indexer:
+        if isinstance(k, slice):
+            backend_indexer.append(k)
+            np_indexer.append(slice(None))
+        else:  # np.ndarray
+            oind, vind = np.unique(k, return_inverse=True)
+            backend_indexer.append(oind)
+            np_indexer.append(vind.reshape(*k.shape))
+
+    backend_indexer = OuterIndexer(tuple(backend_indexer))
+    np_indexer = VectorizedIndexer(tuple(np_indexer))
+
+    if mode == 'outer':
+        return backend_indexer, (np_indexer, )
+
+    shape = tuple(s if isinstance(k, slice) else len(k) for k, s in
+                  zip(backend_indexer.tuple, shape))
+    backend_indexer, np_indexers = _decompose_outer_indexer(
+        backend_indexer, shape, mode)
+    return backend_indexer, (np_indexers[0], np_indexer)
+
+
+def _decompose_outer_indexer(indexer, shape, mode):
+    """
+    Decompose ouer indexer to the successive two indexers, where the
```

Comment:
"ouer" --> "outer"

Also, the phrasing is odd. I would go with "Decompose outer indexer into two indexers" instead, as it is more concise.